crates/miden-tx/src/executor/notes_checker.rs

```diff
@@ -23,17 +26,96 @@ where
+    /// - `failed`: Notes that failed during execution attempts.
     pub async fn check_notes_consumability(
         &self,
         target_account_id: AccountId,
         block_ref: BlockNumber,
         input_notes: InputNotes<InputNote>,
         tx_args: TransactionArgs,
     ) -> Result<NoteConsumptionInfo, TransactionExecutorError> {
-        self.0
-            .try_execute_notes(target_account_id, block_ref, input_notes, tx_args)
+        // Ensure well-known notes are ordered first.
+        let mut notes = input_notes.into_vec();
+        notes.sort_by_key(|note| WellKnownNote::from_note(note.note()).is_none());
+
+        // Attempt to find an executable set of notes.
+        self.find_executable_notes_by_elimination(target_account_id, block_ref, notes, tx_args)
             .await
     }
+
+    /// Finds a set of executable notes and eliminates failed notes from the list in the process.
+    ///
+    /// The result contains some combination of the input notes partitioned by whether they
+    /// succeeded or failed to execute.
+    async fn find_executable_notes_by_elimination(
+        &self,
+        target_account_id: AccountId,
+        block_ref: BlockNumber,
+        notes: Vec<InputNote>,
+        tx_args: TransactionArgs,
+    ) -> Result<NoteConsumptionInfo, TransactionExecutorError> {
+        let mut candidate_notes = notes;
+        let mut failed_notes = Vec::new();
+
+        // Attempt to execute notes in a loop. Reduce the set of notes based on failures until
+        // either a set of notes executes without failure or the set of notes cannot be
+        // further reduced.
+        loop {
+            // Execute the candidate notes.
+            let execution_result = self
+                .0
+                .try_execute_notes(
+                    target_account_id,
+                    block_ref,
+                    InputNotes::<InputNote>::new_unchecked(candidate_notes.clone()),
+                    &tx_args,
+                )
+                .await?;
+            let successful_count = execution_result.successful.len();
+
+            // Determine the next step based on successful and/or failed notes.
+            if successful_count == candidate_notes.len() {
+                // A full set of successful notes has been found.
+                return Ok(NoteConsumptionInfo::new(execution_result.successful, failed_notes));
+            } else if successful_count == 0 {
+                // All notes failed, make no further attempts.
+                failed_notes.extend(execution_result.failed);
+                return Ok(NoteConsumptionInfo::new(Vec::new(), failed_notes));
+            } else {
+                // Some notes succeeded and some failed.
+
+                // Remove the failed notes from the candidate list.
+                let failed_range =
+                    successful_count..successful_count + execution_result.failed.len();
+                candidate_notes.drain(failed_range);
```

Comment:
Looking at how we use `try_execute_notes()` I'm wondering if it actually needs to return `NoteConsumptionInfo` at all. What we really care about is the index of the failing note - and so, maybe we could change it to be something like:
```rust
pub(crate) async fn try_execute_notes(
    &self,
    account_id: AccountId,
    block_ref: BlockNumber,
    notes: InputNotes<InputNote>,
    tx_args: TransactionArgs,
) -> Result<Option<usize>, TransactionExecutorError>
```
And the mapping of the returned value to the 4 cases from the previous comment could look as follows:

1. Returns an error - same as now.
2. Returns `Some(failed_note_idx)`.
3. Returns an error - same as now.
4. Returns `None`.